dipy/align/_public.py

```diff
@@ -0,0 +1,833 @@
+        a string containing the full path to a nifti file.
+
+    static_affine, moving_affine: 4D arrays
+
+    starting_affine : in case this is needed.
+    """
+    static, static_affine = read_img_arr_or_path(static,
+                                                      affine=static_affine)
+    moving, moving_affine = read_img_arr_or_path(moving,
+                                                      affine=moving_affine)
+    if starting_affine is None:
+        starting_affine = np.eye(4)
+
+    return static, static_affine, moving, moving_affine, starting_affine
+
+
+def syn_registration(moving, static,
+                     moving_affine=None,
+                     static_affine=None,
+                     step_length=0.25,
+                     metric='CC',
+                     dim=3,
+                     level_iters=None,
+                     sigma_diff=2.0,
+                     radius=4,
+                     prealign=None):
+    """Register a 2D/3D source image (moving) to a 2D/3D target image (static).
+
+    Parameters
+    ----------
+    moving, static : array or nib.Nifti1Image or str.
+        Either as a 2D/3D array or as a nifti image object, or as
+        a string containing the full path to a nifti file.
+    moving_affine, static_affine : 4x4 array, optional.
+        Must be provided for `data` provided as an array. If provided together
+        with Nifti1Image or str `data`, this input will over-ride the affine
+        that is stored in the `data` input. Default: use the affine stored
+        in `data`.
+    metric : string, optional
+        The metric to be optimized. One of `CC`, `EM`, `SSD`,
+        Default: 'CC' => CCMetric.
+    dim: int (either 2 or 3), optional
+       The dimensions of the image domain. Default: 3
+    level_iters : list of int, optional
+        the number of iterations at each level of the Gaussian Pyramid (the
+        length of the list defines the number of pyramid levels to be
+        used). Default: [10, 10, 5].
+    sigma_diff, radius : float
+        Parameters for initialization of the metric.
+
+    Returns
+    -------
+    warped_moving : ndarray
+        The data in `moving`, warped towards the `static` data.
+    forward : ndarray (..., 3)
+        The vector field describing the forward warping from the source to the
+        target.
+    backward : ndarray (..., 3)
+        The vector field describing the backward warping from the target to the
+        source.
+    """
+    if level_iters is None:
+        level_iters = [10, 10, 5]
+
+    static, static_affine, moving, moving_affine, _ = \
+        _handle_pipeline_inputs(moving, static,
+                                moving_affine=moving_affine,
+                                static_affine=static_affine,
+                                starting_affine=None)
+
+    use_metric = syn_metric_dict[metric](dim, sigma_diff=sigma_diff,
+                                         radius=radius)
```

Comment:
Here, the EM metric will crash since the parameters are different. There are 3 different initialization:
- `CCMetric(dim, sigma_diff=2.0, radius=4)`
- `EMMetric(dim, smooth=1.0, inner_iter=5, q_levels=256, double_gradient=True, step_type='gauss_newton')`
- `SSDMetric(dim, smooth=4, inner_iter=10, step_type='demons')`